processing/src/main/java/io/druid/query/monomorphicprocessing/SpecializationService.java

```diff
@@ -346,18 +349,33 @@ public void run()
           // PerPrototypeClassState.specializationStates. But it might be that nobody ever hits even the current
           // maxSpecializations limit, so implementing cache eviction is an unnecessary complexity.
           specialized = perPrototypeClassState.prototypeClass.newInstance();
-          LOG.warn(
-              "SpecializationService couldn't make more than [%d] specializations. "
-              + "Not doing specialization for runtime shape[%s] and class remapping[%s], using the prototype class[%s]",
-              maxSpecializations,
-              specializationId.runtimeShape,
-              specializationId.classRemapping,
-              perPrototypeClassState.prototypeClass
-          );
+          if (!maxSpecializationsWarningEmitted.get() && maxSpecializationsWarningEmitted.compareAndSet(false, true)) {
+            LOG.warn(
```

Comment:
Current assumption is that nobody will ever hit this, we don't want to miss such event if debug logging is off.